language/complexity-check-plugin/src/complexity-check.ts

```diff
@@ -0,0 +1,261 @@
+import {
+  Diagnostic,
+  DiagnosticSeverity,
+  Range,
+  TextDocument,
+} from "vscode-languageserver-types";
+import { resolveDataRefs } from "@player-ui/player";
+import {
+  getProperty,
+  isPropertyNode,
+  AssetASTNode,
+  type ASTNode,
+  type ViewASTNode,
+  type PlayerLanguageService,
+  type PlayerLanguageServicePlugin,
+  type ASTVisitor,
+} from "@player-tools/json-language-service";
+
+const makeRange = (
+  start: number,
+  end: number,
+  document: TextDocument
+): Range => {
+  return {
+    start: document.positionAt(start),
+    end: document.positionAt(end),
+  };
+};
+
+export interface ComplexityCheckConfig {
+  /** Cutoff for content to be acceptable */
+  maxAcceptableComplexity: number;
+  /**
+   * If set, any score above this number but below `maxAcceptableComplexity` will be logged as a warning
+   */
+  maxWarningLevel?: number;
+  /** If set, maps complexity based on asset or view type */
+  assetComplexity?: Record<string, number>;
```

Comment:
We might want to consider accepting `any` as a key so we can assign weights for assets that are registered with more than just a `type`, as those might have more complexity than the base asset is, i.e. if `{ type: 'asset', role: 'enhanced' }` accepts another asset slot than just `{ type: 'asset' }

Reply:
Thanks for this callout - agreed that including other slots than just type to measure asset complexity would provide additional data points, but at this time we're going to keep it simple just to get this out. I've added in TODO comments so that we don't forget about this.